qiskit/algorithms/gradients/reverse_gradient/reverse_gradient.py

```diff
@@ -0,0 +1,203 @@
+        metadata = []
+
+        for i in range(num_gradients):
+            # temporary variables for easier access
+            circuit = circuits[i]
+            parameters = parameter_sets[i]
+            observable = observables[i]
+            values = parameter_values[i]
+
+            # the metadata only contains the parameters as there are no run configs here
+            metadata.append(
+                {
+                    "parameters": [p for p in circuits[i].parameters if p in parameter_sets[i]],
+                    "derivative_type": self.derivative_type,
+                }
+            )
+
+            # keep track of the parameter order of the circuit, as the circuit splitting might
+            # produce a list of unitaries in a different order
+            original_parameter_order = [p for p in circuit.parameters if p in parameters]
+
+            # split the circuit and generate lists of unitaries [U_1, U_2, ...] and
+            # parameters [p_1, p_2, ...] in these unitaries
+            unitaries, paramlist = split(circuit, parameters=parameters)
+
+            parameter_binds = dict(zip(circuit.parameters, values))
+            bound_circuit = bind(circuit, parameter_binds)
+
+            # initialize state variables -- we use the same naming as in the paper
+            phi = Statevector(bound_circuit)
+            lam = _evolve_by_operator(observable, phi)
+
+            # store gradients in a dictionary to return them in the correct order
+            grads = {param: 0j for param in original_parameter_order}
+
+            num_parameters = len(unitaries)
+            for j in reversed(range(num_parameters)):
+                unitary_j = unitaries[j]
+
+                # We currently only support gates with a single parameter -- which is reflected
+                # in self.SUPPORTED_GATES -- but generally we could also support gates with multiple
+                # parameters per gate
+                parameter_j = paramlist[j][0]
+
+                # get the analytic gradient d U_j / d p_j and bind the gate
+                deriv = derive_circuit(unitary_j, parameter_j)
+                for _, gate in deriv:
+                    bind(gate, parameter_binds, inplace=True)
+
+                # iterate the state variable
+                unitary_j_dagger = bind(unitary_j, parameter_binds).inverse()
+                phi = phi.evolve(unitary_j_dagger)
+
+                # compute current gradient
+                grad = sum(
+                    coeff * lam.conjugate().data.dot(phi.evolve(gate).data) for coeff, gate in deriv
+                )
+
+                # Compute the full gradient (real and complex parts) as all information is available.
+                # Later, based on the derivative type, cast to real/imag/complex.
+                grads[parameter_j] += grad
+
+                if j > 0:
+                    lam = lam.evolve(unitary_j_dagger)
+
+            gradient = np.array(list(grads.values()))
```

Comment:
is it not possible to directly do `gradient = np.array(grads.values())`?

Reply:
Unfortunately not since that'll create an object-array instead of a float-array:
```
>>> np.array({1:1}.values())
array(dict_values([1]), dtype=object)
```